visionmetrics/key_value_pair/key_value_pair_eval_base.py

```diff
@@ -0,0 +1,160 @@
+                                 f" and a 'metric_args' key for a dictionary of arguments to pass to the initialization function of that metric. Key '{key}' does not satisfy this rule.")
+            if metric_name not in [m.value for m in SupportedKeyWiseMetric]:
+                raise ValueError(f"Metric '{key_metric_map[key]['metric_name']}' is not supported. "
+                                 f"Each key's metric must be among the supported metrics: {', '.join([m.value for m in SupportedKeyWiseMetric])}.")
+
+            # Instantiate metric
+            try:
+                self.key_evaluator_map[key] = eval(f"{metric_name}(**metric_args)")
+            except Exception as e:
+                raise ValueError(f"Encountered error '{repr(e)}' when instantiating metric '{metric_name}' for key '{key}' with arguments '{metric_args}'.")
+
+    def update(self, predictions, targets):
+        """
+        Updates metrics for each key using all samples in predictions and targets.
+        Both predictions and targets should be dictionaries of the form {'<key>': <value>}, where <value> is in the format expected for the respective metric for that key.
+        Each sample in predictions and targets must have the same keys (though they do not have to have all the keys in the dataset).
+        """
+        for key in self.key_evaluator_map:
+            metric = self.key_evaluator_map[key]
+            metric_name = self.key_metric_map[key]["metric_name"]
+            key_predictions = []
+            key_targets = []
+            for prediction, target in zip(predictions, targets):
+                # Use the key trace to traverse the prediction and target objects to get the values for these keys
+                try:
+                    key_prediction = reduce(operator.getitem, self.key_metric_map[key]["key_trace"], prediction)
+                except KeyError:
+                    logger.debug(f"No prediction exists in this sample for key '{key}'.")
+                    continue
+                try:
+                    key_target = reduce(operator.getitem, self.key_metric_map[key]["key_trace"], target)
+                except KeyError:
+                    logger.debug(f"No target exists in this sample for key '{key}'.")
+                    continue
+
+                # Construct expected evaluation metric update format for the current key
+                try:
+                    preprocessor = self.key_metric_map[key]["preprocessor"]
+                    key_prediction_formatted, key_target_formatted = preprocessor(key_prediction, key_target)
+                    key_predictions.append(key_prediction_formatted)
+                    key_targets.append(key_target_formatted)
+                except ValueError as e:
+                    logger.debug(f"Encountered error {repr(e)} when preprocessing prediction '{key_prediction}' and target '{key_target}' for key '{key}' to the metric's expected format.")
```

Comment:
Agreed! I originally implemented it this way in case we wanted to run the metric even in cases where the output values might be malformed, but on second thought, since we expect the outputs to be well-formed 100% of the time, it would make more sense just to throw an error.